pkg/cli/images_sign.go

```diff
@@ -0,0 +1,165 @@
+	acornsign "github.com/acorn-io/runtime/pkg/cosign"
+	"github.com/acorn-io/runtime/pkg/images"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/pterm/pterm"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/generate"
+	"github.com/sigstore/cosign/v2/pkg/cosign"
+	"github.com/sigstore/cosign/v2/pkg/cosign/remote"
+	"github.com/sigstore/cosign/v2/pkg/oci/mutate"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"github.com/sigstore/cosign/v2/pkg/oci/static"
+	"github.com/sigstore/cosign/v2/pkg/signature"
+	sigsig "github.com/sigstore/sigstore/pkg/signature"
+	"github.com/spf13/cobra"
+)
+
+func NewImageSign(c CommandContext) *cobra.Command {
+	cmd := cli.Command(&ImageSign{client: c.ClientFactory}, cobra.Command{
+		Use:               "sign IMAGE_NAME [flags]",
+		Example:           `acorn image sign my-image --key ./my-key`,
+		SilenceUsage:      true,
+		Short:             "Sign an Image",
+		ValidArgsFunction: newCompletion(c.ClientFactory, imagesCompletion(true)).complete,
+		Args:              cobra.ExactArgs(1),
+	})
+	_ = cmd.MarkFlagFilename("key")
+	return cmd
+}
+
+type ImageSign struct {
+	client      ClientFactory
+	Key         string            `usage:"Key to use for signing" short:"k" local:"true" default:"./cosign.key"`
+	Annotations map[string]string `usage:"Annotations to add to the signature" short:"a" local:"true"`
+	Push        bool              `usage:"Push the signature to the signature repository" short:"p" local:"true" default:"true"`
+}
+
+func (a *ImageSign) Run(cmd *cobra.Command, args []string) error {
+	c, err := a.client.CreateDefault()
+	if err != nil {
+		return err
+	}
+
+	targetName := args[0]
+	targetDigest := ""
+
+	img, tag, err := client.FindImage(cmd.Context(), c, args[0])
+	if err != nil && !errors.As(err, &images.ErrImageNotFound{}) {
+		return err
+	}
+
+	if err == nil && tag == "" {
+		return fmt.Errorf("Signing a local image without specifying the repository is not supported")
+	} else if tag != "" {
+		targetName = tag
+		targetDigest = img.Digest
+	}
+
+	ref, err := name.ParseReference(targetName)
+	if err != nil {
+		return err
+	}
+
+	if targetDigest == "" {
+		targetDigest, err = acornsign.SimpleDigest(ref)
+		if err != nil {
+			return err
+		}
+	}
+
+	target := ref.Context().Digest(targetDigest)
+
+	if a.Key == "" {
+		return fmt.Errorf("key is required")
+	}
+	pterm.Info.Printf("Signing Image %s (digest: %s) using key %s\n", targetName, targetDigest, a.Key)
+
+	pass, err := generate.GetPass(false)
+	if err != nil {
+		return err
+	}
+
+	pf := func(_ bool) ([]byte, error) {
+		return pass, nil
+	}
+
+	// Get a sigSigner-verifier from a private key and if the key type is not supported, try to import it first
+	var sigSigner sigsig.SignerVerifier
+	sigSigner, err = signature.SignerVerifierFromKeyRef(cmd.Context(), a.Key, pf) // TODO(iwilltry42): use our own style password prompt
+	if err != nil {
+		if !strings.Contains(err.Error(), "unsupported pem type") {
```

Comment:
Is there any way we can cast the error to a specific error type and check that way instead of checking the error string? Ordinarily I am not against this approach if it is in our own code, but since this is coming from a library, we don't know when they might change the wording of the error message for some reason.

Reply:
Unfortunately they don't have a specific error type for this: https://github.com/sigstore/cosign/blob/main/pkg/cosign/keys.go#L216